### PR TITLE
Extend tests for rest of receive module

### PR DESCRIPTION
**Pull request by benalleng**
This expands the mutant  coverage for the rest of the receive module and expands the mutant job coverage over the entire receive module.

Unfortunately the `SHARED_CONTEXT` cannot reasonably be moved into test_utils as the external nature of this struct would then cause errors as the type would be of `payjoin::SessionContext` not `SessionContext`

### CODE REVIEW

**Comment by benalleng** [comment]
Currently marking as draft after looking at the maximum expiry test, I am not satisfied with the syntax there, it has some code smell I want to take another pass at fixing.

File: payjoin/src/receive/v2/mod.rs [hunk 1]
```diff
@@ -707,9 +708,30 @@ mod test {
...
+            session.unwrap().context.expiry.duration_since(SystemTime::now()).unwrap().as_secs();
+        let maximum_expiry = Duration::from_secs(86400);
+        if let Some(expected_expiry) = SystemTime::checked_add(&SystemTime::now(), maximum_expiry) {
+            assert_eq!(TWENTY_FOUR_HOURS_DEFAULT_EXPIRY, maximum_expiry);
+            assert_eq!(
+                session_expiry,
+                expected_expiry.duration_since(SystemTime::now()).unwrap().as_secs()
+            );
+        }
+    }
```
Comment (DanGould) [inline payjoin/src/receive/v2/mod.rs:729]: This test does not pass in my environment, perhaps because some time may pass between calls of SystemTime::now().

```
thread 'receive::v2::test::maximum_expiry' panicked at payjoin/src/receive/v2/mod.rs:724:13:
assertion `left == right` failed
  left: 86399
 right: 86400
```

I think we can use #[tokio::test(start_paused = true)] to pause time so the test is more consistent.

Next, checked_add call is a bit odd if it's using &self. Prefer `SystemTime::now().checked_add(maximum_expiry)`

When I tested this I also got an issue compiling with the [start_paused](https://docs.rs/tokio/1.12.0/tokio/runtime/struct.Builder.html#method.start_paused) change which made me bump tokio. Seems like 1.38.1 is 1.63 MSRV. Turned out the bump wasn't necessary but the `ttokio/est-util` feature was. Not a bad time to bump to 1.38.1 in a separate commit, tho.

Lastly I'm not sure why thi stest is named maximum expiry. Isn't it default? Couldn't it be set to be longer?

File: payjoin/src/receive/multiparty/mod.rs [hunk 2]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
...
+    use payjoin_test_utils::{BoxError, PARSED_ORIGINAL_PSBT};
+
+    use super::{v1, v2, FinalizedProposal, UncheckedProposalBuilder};
+    use crate::receive::optional_parameters::Params;
+    use crate::receive::v2::test::SHARED_CONTEXT;
+
+    fn multiparty_proposal_from_test_vector() -> v1::UncheckedProposal {
+        let pairs = url::form_urlencoded::parse("v=2&optimisticmerge=true".as_bytes());
+        let params =
+            Params::from_query_pairs(pairs, &[2]).expect("Could not parse from query pairs");
```
Comment (DanGould) [inline payjoin/src/receive/multiparty/mod.rs:266]: I stared at this `&[2]` for a while wondering what it was. Makes me wonder if we ought to have a `Version` enum.

File: payjoin/src/receive/multiparty/mod.rs [hunk 3]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
 
     pub fn v2(&self) -> &[v2::UncheckedProposal] { &self.v2_proposals }
 }
+
+#[cfg(test)]
+mod test {
+
+    use std::any::{Any, TypeId};
+
+    use payjoin_test_utils::{BoxError, PARSED_ORIGINAL_PSBT};
```
Comment (DanGould) [inline payjoin/src/receive/multiparty/mod.rs:257]: Doesn't seem like a big reason to import and use `BoxError` or results if the functions return `Ok(())` and the `?` isn't used anywhere.

**Review by DanGould — COMMENTED** [review]
I'd want @0xBEEFCAF3 to check out the multiparty before ACK.

I left some comments on one issue that for sure needs fixing and another that'd be nice.

Crushing deez receive mutants 🥾

File: payjoin/src/receive/multiparty/mod.rs [hunk 4]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
...
+            v1: multiparty_proposal_from_test_vector(),
+            context: SHARED_CONTEXT.clone(),
+        };
+        let mut multiparty = UncheckedProposalBuilder::new();
+        multiparty.add(proposal_one).expect("Could not add proposal to multiparty");
+        multiparty.add(proposal_two).expect("Could not add proposal to multiparty");
+        let unchecked_proposal = multiparty.build();
+        assert!(
+            unchecked_proposal.expect("Could not build multiparty proposal").contexts.len() == 2
+        );
```
Comment (0xBEEFCAF3) [inline payjoin/src/receive/multiparty/mod.rs:286]: This is working as expected but I do wonder if contexts here should be len == 1. Since both the hpke context and the proposal is the same it would make sense to me that  `UncheckedProposalBuilder.add()` would disallow the caller to add duplicate sessions. I still think the ns1r session would succeed bc only one PSBT ends up on the directory for this hpke session but it still seems wrong and error prone. This is outside the scope of this PR.

File: payjoin/src/receive/multiparty/mod.rs [hunk 5]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
...
+            v1: multiparty_proposal_from_test_vector(),
+            context: SHARED_CONTEXT.clone(),
+        };
+        let mut finalized_multiparty = FinalizedProposal::new();
+        finalized_multiparty
+            .add(proposal_one.clone())
+            .expect("Could not add propsal to multiparty");
+
+        finalized_multiparty.add(proposal_two).expect("Could not add propsal to multiparty");
+        assert_eq!(finalized_multiparty.v2()[0].type_id(), TypeId::of::<v2::UncheckedProposal>());
```
Comment (0xBEEFCAF3) [inline payjoin/src/receive/multiparty/mod.rs:307]: Mind as well check the second proposal as well. 
```suggestion
        assert_eq!(finalized_multiparty.v2()[0].type_id(), TypeId::of::<v2::UncheckedProposal>());
        assert_eq!(finalized_multiparty.v2()[1].type_id(), TypeId::of::<v2::UncheckedProposal>());

```

File: payjoin/src/receive/multiparty/mod.rs [hunk 6]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
...
+            context: SHARED_CONTEXT.clone(),
+        };
+        let mut multiparty = UncheckedProposalBuilder::new();
+        multiparty.add(proposal_one).expect("Could not add proposal to multiparty");
+        multiparty.add(proposal_two).expect("Could not add proposal to multiparty");
+        let unchecked_proposal = multiparty.build();
+        assert!(
+            unchecked_proposal.expect("Could not build multiparty proposal").contexts.len() == 2
+        );
+        Ok(())
```
Comment (0xBEEFCAF3) [inline payjoin/src/receive/multiparty/mod.rs:287]: It any case we should also have a test where we add different PSBTs with different sessions as this is the happy path. This can also be a follow up!

File: payjoin/src/receive/multiparty/mod.rs [hunk 7]
```diff
@@ -248,3 +248,68 @@ impl FinalizedProposal {
...
+            v1: multiparty_proposal_from_test_vector(),
+            context: SHARED_CONTEXT.clone(),
+        };
+        let proposal_two = v2::UncheckedProposal {
+            v1: multiparty_proposal_from_test_vector(),
+            context: SHARED_CONTEXT.clone(),
+        };
+        let mut multiparty = UncheckedProposalBuilder::new();
+        multiparty.add(proposal_one).expect("Could not add proposal to multiparty");
+        multiparty.add(proposal_two).expect("Could not add proposal to multiparty");
```
Comment (0xBEEFCAF3) [inline payjoin/src/receive/multiparty/mod.rs:282]: Can we use `?` here instead of expect?

**Review by 0xBEEFCAF3 — APPROVED** [review]
Ack on the multi party front. 47d075fecaa61abbeecbf7d34fad36d0d0780c4c
Just had a couple small comments, no show stoppers. 

Its great to see some test coverage on the proposal builders. The tests added are actually testing what I would classify as an edge case where we are building multi party proposal with duplicate inner proposals. In a follow up PR we should fix this and then follow up with some tests on the happy path as well (unique proposals with different hpke sessions).

Thanks for adding these :rocket:

File: payjoin/src/receive/multiparty/mod.rs (hunk 4 above)
Comment (0xBEEFCAF3) [reply payjoin/src/receive/multiparty/mod.rs:286]: Issue created: https://github.com/payjoin/rust-payjoin/issues/640

**Comment by benalleng** [comment]
> In a follow up PR we should fix this and then follow up with some tests on the happy path as well (unique proposals with different hpke sessions).

Yeah that's definitely something I am trying to keep in mind, Mutant catches rarely seem to be easily tested with the happy path when they can't be caught with a simple assert.

File: payjoin/src/receive/v2/mod.rs (hunk 1 above)
Comment (benalleng) [reply payjoin/src/receive/v2/mod.rs]: Ok! Need to spend more time learning how to use tokio

**Review by DanGould — CHANGES_REQUESTED** [review]
Nice changes, But the feature addition needs to be fixup!ed into the first commit or else the first commit will not compile 

Please also change the second commit to change all `tokio = <version>` definitions to match the update.

**Comment by DanGould** [comment]
I'm still getting the local failure despite CI passing. Going to have to troubleshoot hmm

```console

running 1 test
test receive::v2::test::default_expiry ... FAILED

successes:

successes:

failures:

---- receive::v2::test::default_expiry stdout ----

thread 'receive::v2::test::default_expiry' panicked at payjoin/src/receive/v2/mod.rs:725:13:
assertion `left == right` failed
  left: 86399
 right: 86400
```

**Comment by benalleng** [comment]
@DanGould give this recent one a shot, I tried just initializing `SystemTime::now()` once and call it several times so maybe this will be more consistent

**Comment by DanGould** [comment]
This works! However it no longer depends on tokio::test(start_paused = true), so please one more change to switch back to #[test] and shut the tokio feature off. The tokio should stay, good to bring deps up to date.

**Review by DanGould — APPROVED** [review]
~~ACK 180f3084d2666fff6a6c0500063bb28266e6a80b~~

whoops, that's the rebase on master

ACK 9be9dc25acddd0ac41fb2c21d68685b3e9248edc
